### PR TITLE
pkg(both): a separate library for DML row change

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflow/issues/3242

### What is changed and how it works?

provide a representation for DML (row change), and some conversion of it. Also fix some wrong behaviour of original code (such as we should decide generated column from target table structure)

next PR I'll let DM adopt this library, and future plans are
- CDC adopt this library
- propose a DML worker which include reduce/causality/executeSQL functionality

### Check List 

Tests 

 - Unit test
 - after adopted this PR, DM can pass CI https://ci2.pingcap.net/blue/organizations/jenkins/dm_ghpr_integration_test/detail/dm_ghpr_integration_test/2843/pipeline

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

Related changes

### Release note 

```release-note
 `None`.
```

### CODE REVIEW

File: pkg/sqlmodel/row_change.go [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+	if downstreamTableInfo != nil {
+		ret.targetTableInfo = downstreamTableInfo
+	} else {
+		ret.targetTableInfo = sourceTableInfo
+	}
+
+	if tiCtx != nil {
+		ret.tiSessionCtx = tiCtx
+	} else {
+		ret.tiSessionCtx = utils.NewSessionCtx(nil)
```
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/sqlmodel/row_change.go]: changed to use a global variable

File: dm/pkg/utils/common.go [hunk 2]
```diff
@@ -325,6 +325,8 @@ func (se *session) GetBuiltinFunctionUsage() map[string]uint32 {
 // NewSessionCtx return a session context with specified session variables.
 func NewSessionCtx(vars map[string]string) sessionctx.Context {
 	variables := variable.NewSessionVars()
+	variables.Rng.SetSeed1(0)
+	variables.Rng.SetSeed2(0)
```
[the remark this answers is not in the document]
Comment (lance6716) [reply dm/pkg/utils/common.go]: Removed now. Originally I'm use different session context instance for each RowChange, their random seed fields is different so can't pass DeepEquals

**Comment by lance6716** [comment]
I think this library can provide different Causality function like RowCausality, TableCausality :thinking: 

```
type CausalityKeyFunc func (change *RowChange) []string
```

Will add it in future PR because I want to reduce divergence with https://github.com/pingcap/tiflow/pull/4313

File: pkg/sqlmodel/multivalue.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+import (
+	"strings"
+
+	"go.uber.org/zap"
+
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/pkg/quotes"
+)
+
+// SameTypeTargetAndColumns check whether two row changes have same type, target and columns, so they can be merged to a
```
Comment (overvenus) [inline pkg/sqlmodel/multivalue.go:25]: Please keep line width within 80 chars.

File: pkg/sqlmodel/row_change.go [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+func (r *RowChange) lazyInitIdentityInfo() {
+	if r.identityInfo != nil {
+		return
+	}
+
+	r.identityInfo = schema.GetDownStreamTI(r.targetTableInfo, r.sourceTableInfo)
+}
+
+func getColsAndValuesOfIdx(columns []*timodel.ColumnInfo, indexColumns *timodel.IndexInfo, data []interface{}) ([]*timodel.ColumnInfo, []interface{}) {
```
Comment (overvenus) [inline pkg/sqlmodel/row_change.go:180]: Please keep line width within 80 chars.

Comment (overvenus) [reply pkg/sqlmodel/row_change.go:180]: 1 table = 4 space.

File: pkg/sqlmodel/row_change.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+			buf.WriteString(" IS ?")
+		} else {
+			buf.WriteString(" = ?")
+		}
+	}
+	return whereValues
+}
+
+// valuesHolder gens values holder like (?,?,?).
+// n must be greater or equal than 1, or the function will panic.
```
Comment (amyangfei) [inline pkg/sqlmodel/row_change.go:250]: seems n==0 also works, returns `()`

**Review by glorv — COMMENTED** [review]
LGTM

**Comment by lance6716** [comment]
ping @amyangfei @liuzix @GMHDBJD

**Review by amyangfei — APPROVED** [review]

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Comment by lance6716** [comment]
/hold

test failed

**Comment by lance6716** [comment]
after DPanic takes effect, I found a problem https://github.com/pingcap/tiflow/pull/4376/commits/f2c4bcb04e807592ee14f402200f15c93d8583c0

PTAL @GMHDBJD and `/merge` if looks good

**Pull request merged**